plugins/inputs/sqlserver/sqlserver.go

```diff
@@ -423,6 +464,96 @@ func (s *SQLServer) Init() error {
 	return nil
 }
 
+// Get Token Provider by loading cached token or refreshed token
+func getTokenProvider() (func() (string, error), error) {
+	var tokenString string
+
+	// load token
+	muCacheLock.RLock()
+	token, err := loadToken()
+	muCacheLock.RUnlock()
+
+	// if there's error while loading token or found an expired token, refresh token and save it
+	if err != nil || token.IsExpired() {
+		// refresh token within a write-lock
+		muCacheLock.Lock()
+
+		// load token again, in case it's been refreshed by another thread
+		token, err = loadToken()
+
+		// check loaded token's error/validity, then refresh/save token
+		if err != nil || token.IsExpired() {
+			// get new token
+			spt, err := refreshToken()
+			if err != nil {
+				return nil, err
+			}
+
+			// use the refreshed token
+			tokenString = spt.OAuthToken()
+		} else {
+			// use locally cached token
+			tokenString = token.OAuthToken()
+		}
+
+		muCacheLock.Unlock()
+	} else {
+		// use locally cached token
+		tokenString = token.OAuthToken()
+	}
+
+	// return acquired token
+	return func() (string, error) {
+		return tokenString, nil
+	}, nil
+}
+
+// Load token from in-mem cache
+func loadToken() (*adal.Token, error) {
+	if adalToken == nil {
+		return nil, fmt.Errorf("Token is nil or Failed to load existing token")
+	}
+
+	return adalToken, nil
+}
```

Comment:
I don't get the intention behind this function... You could check for the token being `nil` on the caller side or am I missing something...

Reply:
This was added while looking for different options of caching the token (for example, disk based), so would be useful if we use a different caching mechanism. I'd prefer to keep it structured this way, unless you have a strong opinion on it.

Reply:
Fine with me, maybe you could add a comment about this reasoning to the code?

Reply:
Btw: Error messages should start lowercase... ;-P

Reply:
Added a comment to the method and the error message with lowercase was already fixed in the previous commit (above screen shot is "outdated" :-) )